src/box/iproto.cc

```diff
@@ -686,6 +692,15 @@ iproto_connection_try_to_start_destroy(struct iproto_connection *con)
 		con->state = IPROTO_CONNECTION_PENDING_DESTROY;
 		return;
 	}
+	/*
+	 * If shutdown process started, part of iproto_connection struct
+	 * used there, so freeing of connection will lead to UB. Let shutdown
+	 * process destroy the connection.
+	 */
+	if (con->iproto_thread->is_shutdown_active &&
+	    !rlist_empty(&con->in_all_list))
```

Comment:
That is why we have connection state instead of flags. You tried to form a new state from some flags and rlists, but wouldn't it be simpler to add a new state? And why couldn't you add that check to `iproto_connection_is_idle()`? The connection does not get PENDING_DESTROY state in your code but it should. Shouldn't it? Try_to_start_destroy just left it unchanged, which looks wrong.